http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/RequestAdapter.java

```diff
@@ -17,27 +17,56 @@
 
 import io.netty.handler.codec.http.DefaultHttpHeaders;
 import io.netty.handler.codec.http.DefaultHttpRequest;
+import io.netty.handler.codec.http.HttpHeaderNames;
 import io.netty.handler.codec.http.HttpHeaders;
 import io.netty.handler.codec.http.HttpMethod;
 import io.netty.handler.codec.http.HttpRequest;
 import io.netty.handler.codec.http.HttpVersion;
+import java.util.Arrays;
+import java.util.List;
 import software.amazon.awssdk.annotations.SdkInternalApi;
 import software.amazon.awssdk.http.SdkHttpMethod;
 import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.utils.http.SdkHttpUtils;
 
 @SdkInternalApi
 public final class RequestAdapter {
 
+    private static final List<String> IGNORE_HEADERS = Arrays.asList(HttpHeaderNames.HOST.toString());
+
     public HttpRequest adapt(SdkHttpRequest sdkRequest) {
         HttpMethod method = toNettyHttpMethod(sdkRequest.method());
         HttpHeaders headers = new DefaultHttpHeaders();
         String uri = sdkRequest.getUri().toString();
         DefaultHttpRequest request = new DefaultHttpRequest(HttpVersion.HTTP_1_1, method, uri, headers);
-        sdkRequest.headers().forEach(request.headers()::add);
+        addHeadersToRequest(request, sdkRequest);
```

Comment:
`getUri()` in this case returns the absolute URI. Since we're setting the `Host` header now, let's change  `uri` to just be the absolute path of the request.

Reply:
That would make the behavior inconsistent with the Apache implementation - it only rewrites the URI to work around handling of consecutive `/` characters in the path.

Reply:
We shouldn't be rewriting anything, we can just switch to using the path part of the URI now that we're setting the `Host` header separately. For that we can use `sdkRequest.getUri().getPath()`. The [docs](https://netty.io/4.1/api/io/netty/handler/codec/http/DefaultHttpRequest.html#DefaultHttpRequest-io.netty.handler.codec.http.HttpVersion-io.netty.handler.codec.http.HttpMethod-java.lang.String-) for Netty's `DefaultHttpRequest` state that the `uri` parameter is either the request URI or the path. This should bring the behavior in line Apache.

Without this change, the request through Netty will be something like the following, where the request line contains the absolute URI, but also has the `Host` header

```
GET https://s3.us-west-2.amazonaws.com/ HTTP/1.1
amz-sdk-invocation-id: 7aee6184-1868-578e-4c55-0ce056519ac2
amz-sdk-retry: 0/0/500
Authorization: AWS4-HMAC-SHA256 Credential=AKIAISLBO3T24I4JYG4Q/20191024/us-west-2/s3/aws4_request, SignedHeaders=amz-sdk-invocation-id;amz-sdk-retry;host;x-amz-content-sha256;x-amz-date, Signature=5ca4e8bfcc44b68f865860de5e5cca0c9cc1ad9e85849d3c8f78bd0a73795d6a
Host: s3.us-west-2.amazonaws.com
User-Agent: aws-sdk-java/2.9.11 Mac_OS_X/10.13.6 OpenJDK_64-Bit_Server_VM/25.222-b10 Java/1.8.0_222 vendor/Amazon.com_Inc. io/async http/NettyNio
x-amz-content-sha256: UNSIGNED-PAYLOAD
X-Amz-Date: 20191024T003545Z
```

Using Apache it looks something like the following, where the request line only contains the path

```
>> "GET / HTTP/1.1[\r][\n]"
>> "Host: s3.us-west-2.amazonaws.com[\r][\n]"
>> "amz-sdk-invocation-id: 75fd9d45-2ecd-b72d-5a30-12d53df6617e[\r][\n]"
>> "amz-sdk-retry: 0/0/500[\r][\n]"
>> "Authorization: AWS4-HMAC-SHA256 Credential=AKIAISLBO3T24I4JYG4Q/20191024/us-west-2/s3/aws4_request, SignedHeaders=amz-sdk-invocation-id;amz-sdk-retry;host;x-amz-content-sha256;x-amz-date, Signature=fff62d8e6d929faf0843a5bfa95db05b4e178553149c3c646bf789afa810e6de[\r][\n]"
>> "User-Agent: aws-sdk-java/2.9.11 Mac_OS_X/10.13.6 OpenJDK_64-Bit_Server_VM/25.222-b10 Java/1.8.0_222 vendor/Amazon.com_Inc. io/sync http/Apache[\r][\n]"
>> "x-amz-content-sha256: UNSIGNED-PAYLOAD[\r][\n]"
>> "X-Amz-Date: 20191024T001821Z[\r][\n]"
>> "Connection: Keep-Alive[\r][\n]"
>> "[\r][\n]"
```

But the same request with these changes on the Netty client will be

Reply:
ah whoops, my bad. I must not have read through the Apache httpclient code thoroughly enough (and definitely didn't look at the raw request content). I'll work on updating the PR.

Reply:
just a comment for myself - the relevant logic in the Apache implementation to do this is in https://github.com/apache/httpcomponents-client/blob/4.5.9/httpclient/src/main/java/org/apache/http/client/utils/URIUtils.java

Reply:
Okay the build jobs are failing in the HTTP 2 tests (e.g. GoAwayTest) due to an IllegalArgumentException `:scheme must be specified. see https://tools.ietf.org/html/rfc7540#section-8.1.2.3` being thrown, however the internal error handling seems to be a bit poor, to the point that the error message is never printed out and the `CompletableFuture.join()` calls aren't throwing the exception. I had to step through it in a debugger session.

Looks like there's a potential workaround by explicitly setting `ExtensionHeaderNames.SCHEME` if we are on HTTP/2. I'll work on that and update the PR again.